AnaTutorial/src/AnaTutorial.cc

```diff
@@ -0,0 +1,848 @@
+          dphi += TMath::Pi() * 2.;
+
+        float deta = recojeteta - thisjeteta;
+        /// Determine the distance in eta phi space between the reconstructed
+        /// and truth jets
+        dR = sqrt(pow(dphi, 2.) + pow(deta, 2.));
+
+        /// If this truth jet is closer than the previous truth jet, it is
+        /// closer and thus should be considered the truth jet
+        if (dR < reco_jets->get_par() && dR < closestjet)
+        {
+          truthjetid = -9999;
+          truthjetp = jet->get_p();
+          truthjetphi = jet->get_phi();
+          truthjeteta = jet->get_eta();
+          truthjetpt = jet->get_pt();
+          truthjetenergy = jet->get_e();
+          truthjetpx = jet->get_px();
+          truthjetpy = jet->get_py();
+          truthjetpz = jet->get_pz();
+          closestjet = dR;
+        }
+      }
+    }
+    recojettree->Fill();
+  }
+}
+
+/**
+ * This method gets clusters from the EMCal and stores them in a tree. It
+ * also demonstrates how to get trigger emulator information. Clusters from
+ * other containers can be obtained in a similar way (e.g. clusters from
+ * the IHCal, etc.)
+ */
+void AnaTutorial::getEMCalClusters(PHCompositeNode *topNode)
+{
+  /// Get the raw cluster container
+  /// Note: other cluster containers exist as well. Check out the node tree when
+  /// you run a simulation
+  RawClusterContainer *clusters = findNode::getClass<RawClusterContainer>(topNode, "CLUSTER_CEMC");
+
+  if (!clusters)
+  {
+    cout << PHWHERE
+         << "EMCal cluster node is missing, can't collect EMCal clusters"
+         << endl;
+    return;
+  }
+
+  /// Get the global vertex to determine the appropriate pseudorapidity of the clusters
+  GlobalVertexMap *vertexmap = findNode::getClass<GlobalVertexMap>(topNode, "GlobalVertexMap");
+  if (!vertexmap)
+  {
+    cout << "AnaTutorial::getEmcalClusters - Fatal Error - GlobalVertexMap node is missing. Please turn on the do_global flag in the main macro in order to reconstruct the global vertex." << endl;
+    assert(vertexmap);  // force quit
+
+    return;
+  }
+
+  if (vertexmap->empty())
+  {
+    cout << "AnaTutorial::getEmcalClusters - Fatal Error - GlobalVertexMap node is empty. Please turn on the do_global flag in the main macro in order to reconstruct the global vertex." << endl;
+    return;
+  }
+
+  GlobalVertex *vtx = vertexmap->begin()->second;
+  if (vtx == nullptr)
+    return;
+
+  /// Trigger emulator
+  CaloTriggerInfo *trigger = findNode::getClass<CaloTriggerInfo>(topNode, "CaloTriggerInfo");
+  /// Can obtain some trigger information if desired
+  E_4x4 = trigger->get_best_EMCal_4x4_E();
+
+  RawClusterContainer::ConstRange begin_end = clusters->getClusters();
+  RawClusterContainer::ConstIterator clusiter;
+
+  /// Loop over the EMCal clusters
+  for (clusiter = begin_end.first;
+       clusiter != begin_end.second;
+       ++clusiter)
+  {
+    /// Get this cluster
+    RawCluster *cluster = clusiter->second;
```

Comment:
It is somewhat safer to user if we require `cluster` be read only
`const RawCluster *cluster ...`

Reply:
Implemented